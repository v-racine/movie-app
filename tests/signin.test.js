const { Config } = require('../src/config/config');

Config.Get(process.env);

const { AppFactory, store } = require('../src/app');
const request = require('supertest');

describe('sign in', () => {
  const mockPgClient = {};

  const app = AppFactory({
    pgClient: mockPgClient,
  });

  describe('when: a user wants to view the form to signin', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app).get('/signIn').send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return the view of the form', async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe("when: the user attempt to submit the 'sign-in' form BUT enters an incorrect email", () => {
    let rsp;
    let id = '3424';

    let user = {
      id: id,
      username: 'Phillips',
      email: 'test@test.com',
      password: 'asdf',
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementationOnce(() => {
        return new Promise((resolve) => {
          resolve({ rows: [] });
        });
      });

      rsp = await request(app)
        .post('/signIn')
        .send(`username=${user.username}&email=${user.email}&password=${user.password}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Email not found'", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toContain('Email not found');
    });
  });

  describe("when: the user attempt to submit the 'sign-in' form BUT enters an incorrect password", () => {
    let rsp;
    let id = '3424';

    let user = {
      id: id,
      username: 'Phillips',
      email: 'test@test.com',
      password: 'papillon2',
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementationOnce(() => {
        return new Promise((resolve) => {
          resolve({
            rows: [
              {
                username: 'Phillips',
                email: 'test@test.com',
                password:
                  'f0cb357c9960fc49f80c807fbaaf45be391fb69502b22ed29149a7b67e3cb66ffa1cb747e62c7419ff548b0061d1a97d84d8b610da545b94a11c64297e763d8f.e40ff17e1d215c67',
              },
            ],
          });
        });
      });

      rsp = await request(app)
        .post('/signIn')
        .send(`username=${user.username}&email=${user.email}&password=${user.password}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Incorrect password'", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toContain('Incorrect password');
    });
  });

  describe("when: the user successfully submits the 'sign-in' form", () => {
    let rsp;

    let id = '3424';

    let user = {
      id: id,
      username: 'Phillips',
      email: 'test@test.com',
      password: 'papillon',
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementationOnce(() => {
        return new Promise((resolve) => {
          resolve({
            rows: [
              {
                id: id,
                username: 'Phillips',
                password:
                  'f0cb357c9960fc49f80c807fbaaf45be391fb69502b22ed29149a7b67e3cb66ffa1cb747e62c7419ff548b0061d1a97d84d8b610da545b94a11c64297e763d8f.e40ff17e1d215c67',
              },
            ],
          });
        });
      });

      rsp = await request(app)
        .post('/signIn')
        .send(`username=${user.username}&email=${user.email}&password=${user.password}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we redirect to the movies page with a flash message, 'You're signed in!'", async () => {
      expect(mockPgClient.dbQuery).toHaveBeenCalledTimes(1);

      expect(mockPgClient.dbQuery).toHaveBeenNthCalledWith(
        1,
        'SELECT * FROM users WHERE username = $1 AND email = $2',
        user.username,
        user.email,
      );

      const toClient = rsp.get('set-cookie')[0];
      expect(toClient).toBeDefined();

      const status = rsp.status;
      expect(status).toBe(302);
    });
  });
});
