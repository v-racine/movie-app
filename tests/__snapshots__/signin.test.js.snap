// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign in when: a user wants to view the form to signin then: we return the view of the form 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <p>
      </p>
      <b>
        Sign In:
      </b>
      <form action="/signIn"
            method="post"
      >
        <dl>
          <dt>
            <label for="username">
              Enter your username:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="username"
                   name="username"
                   placeholder="name"
            >
          </dd>
          <dt>
            <label for="email">
              Enter your email:
            </label>
          </dt>
          <dd>
            <input type="email"
                   id="email"
                   name="email"
                   placeholder="username@example.com"
            >
          </dd>
          <dt>
            <label for="password">
              Enter your password:
            </label>
          </dt>
          <dd>
            <input type="password"
                   id="password"
                   name="password"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Submit"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
      <p>
        <a href="/signUp">
          Don't have an account? Sign Up
        </a>
      </p>
    </main>
  </body>
</html>
`;
